packages/perennial-oracle/contracts/ChainlinkFeedOracle.sol

```diff
@@ -0,0 +1,157 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity 0.8.15;
+
+import "@chainlink/contracts/src/v0.8/interfaces/AggregatorV2V3Interface.sol";
+import "@openzeppelin/contracts/utils/math/SafeCast.sol";
+import "./interfaces/IOracleProvider.sol";
+import "./types/ChainlinkAggregator.sol";
+
+/**
+ * @title ChainlinkOracle
+ * @notice Chainlink implementation of the IOracle interface.
+ * @dev One instance per Chainlink price feed should be deployed. Multiple products may use the same
+ *      ChainlinkOracle instance if their payoff functions are based on the same underlying oracle.
+ *      This implementation only support non-negative prices.
+ */
+contract ChainlinkFeedOracle is IOracleProvider {
+    error InvalidVersionRequested(uint256 version);
+
+    /// @dev Chainlink feed aggregator address
+    ChainlinkAggregator public immutable aggregator;
+
+    /// @dev Decimal offset used to normalize chainlink price to 18 decimals
+    int256 private immutable _decimalOffset;
+
+    /// @dev First valid phase for this Oracle
+    uint256 private _firstPhaseId;
+
+    /// @dev Mapping of the starting data for each underlying phase
+    Phase[] private _phases;
+
+    struct Phase {
+        uint128 startingVersion;
+        uint128 startingRoundId;
+    }
+
+    /**
+     * @notice Initializes the contract state
+     * @param aggregator_ Chainlink price feed registry
+     */
+    constructor(ChainlinkAggregator aggregator_) {
+        aggregator = aggregator_;
+
+        // phaseId is 1-indexed, skip index 0
+        _phases.push(Phase(uint128(0), uint128(0)));
+        // phaseId is 1-indexed, first phase starts as version 0
+        _phases.push(Phase(uint128(0), uint128(_aggregatorRoundIdToProxyRoundId(1, 1))));
+
+        _decimalOffset = SafeCast.toInt256(10 ** aggregator.decimals());
+
+        _firstPhaseId = aggregator_.phase();
+        // Load the phases array with previous phase values
+        while (_firstPhaseId > _latestPhaseId()) {
+            _phases.push(Phase(_phases[_phases.length - 1].startingVersion, uint128(0)));
+        }
+    }
+
+    /**
+     * @notice Checks for a new price and updates the internal phase annotation state accordingly
+     * @return The current oracle version after sync
+     */
+    function sync() external returns (OracleVersion memory) {
+        // Fetch latest round
+        ChainlinkRound memory round = aggregator.getLatestRound();
+
+        // Update phase annotation when new phase detected
+        while (round.phaseId() > _latestPhaseId()) {
+            _phases.push(
+                Phase(
+                    uint128(aggregator.getRoundCount(_latestPhaseId(), round.timestamp)) +
+                        _phases[_phases.length - 1].startingVersion,
+                    uint128(_aggregatorRoundIdToProxyRoundId(_latestPhaseId() + 1, 1))
```

Comment:
I think you want to just use the `round.roundId` here instead of deriving it -- we may already be a few version in (if the sync is late) and this won't work properly with the timestamp walkback if we assume we start at 1

Reply:
Oops, yeah forgot to fix this with the new walkback logic.